# Table of Contents

0.1 Importing Libraries

0.2 Importing Data

0.3 Exploring Original Df

0.4 Create new columns with profile flags

    0.4.1  income_group
    
    0.4.2  senior_citizen
    
    0.4.3  two_parent_fam
    
    0.4.4  single_parent
      
    0.4.5  baby_parent
    
    0.4.6  pet_owner
    
0.5 Merging temp subset with full df

0.6 Confirming Final Df 

0.7 Exporting Final Df

### 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 0.2 Importing Data

In [2]:
# Identify the file pathway to data files
path = r'C:\Users\CJ\Documents\_CJ-Stuff\Career Foundry\Data Immersion\Ach 4 - Python\2023-03 Instacart Basket Analysis'

In [3]:
# Import data
df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all2.pkl'))

### 0.3 Exploring Original Df

In [4]:
df.shape

(32434489, 39)

In [5]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... date_joined  n_dependants  fam_status  \
0                  1          0  ...  2019-02-17             3     married   
1                  2          0  ...  2019-02-17             3     married   
2                  3          0  ...  2019-02-17             3     married   
3                  4          0  ...  2019-02-17             3     married   
4                  5          0  ...  2019-02-17             3     married   

   income  region  low_activity_flag total_spent_in_order  \
0   40423   South              False                 31.6   
1   40423   South              False                 31.6   
2   40423   South              False                 31.6   
3   40423   South              False                 31.6   
4   40423   South              False                 31.6   

  total_spent_in_lifetime total_items_in_lifetime avg_items_per_order_by_user  
0              375.700012                      59                    5.898438  
1              375.700012                      59                    5.898438  
2              375.700012                      59                    5.898438  
3              375.700012                      59                    5.898438  
4              375.700012                      59                    5.898438  

[5 rows x 39 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434489 entries, 0 to 32434488
Data columns (total 39 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   order_id                     int64         
 1   user_id                      int64         
 2   order_number                 int16         
 3   order_day_of_week            int8          
 4   order_hour_of_day            int8          
 5   days_since_prior_order       float16       
 6   first_order_of_new_customer  bool          
 7   product_id                   int32         
 8   add_to_cart_order            int16         
 9   reordered                    int8          
 10  product_name                 object        
 11  aisle_id                     float16       
 12  department_id                float16       
 13  price                        float32       
 14  product_info_available       bool          
 15  price_available              bool          
 16

In [7]:
# Checking for nulls
df.isnull().sum()

order_id                             0
user_id                              0
order_number                         0
order_day_of_week                    0
order_hour_of_day                    0
days_since_prior_order         2078068
first_order_of_new_customer          0
product_id                           0
add_to_cart_order                    0
reordered                            0
product_name                     30770
aisle_id                         30770
department_id                    30770
price                            35897
product_info_available               0
price_available                      0
price_range                      35897
busiest_day                          0
busiest_days                         0
busiest_period_of_day                0
max_order                            0
loyalty_flag                         0
mean_spending                        0
spender_type                         0
median_days_since_ordering           0
frequent_shopper         

In [8]:
# The null values for days_since_prior_order 
# are flagged by first_order of new customer
df['first_order_of_new_customer'].value_counts()

False    30356421
True      2078068
Name: first_order_of_new_customer, dtype: int64

In [9]:
# The null values for price
# are flagged by price_available
df['price_available'].value_counts()

True     32398592
False       35897
Name: price_available, dtype: int64

In [10]:
# The null values for product_name, aisle_id, and department_id 
# are flagged by product_info_available
df['product_info_available'].value_counts()

True     32403719
False       30770
Name: product_info_available, dtype: int64

All null/NaN values are appropriately flagged.

In [11]:
df.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.243449e+07  3.243449e+07  3.243449e+07       3.243449e+07   
mean   1.710749e+06  1.029372e+05  1.714205e+01       2.738818e+00   
std    9.873007e+05  5.946648e+04  1.753504e+01       2.090049e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.559430e+05  5.142100e+04  5.000000e+00       1.000000e+00   
50%    1.711048e+06  1.026110e+05  1.100000e+01       3.000000e+00   
75%    2.565514e+06  1.543910e+05  2.400000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01       6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.243449e+07              30356421.0  3.243449e+07   
mean        1.342498e+01                     NaN  2.557634e+04   
std         4.246365e+00                     0.0  1.409669e+04   
min         0.000000e+00                     0.0  1.000000e+00   
25%         1.000000e+01                     5.0  1.353000e+04   
50%         1.300000e+01                     8.0  2.525600e+04   
75%         1.600000e+01                    15.0  3.793500e+04   
max         2.300000e+01                    30.0  4.968800e+04   

       add_to_cart_order     reordered    aisle_id  ...     max_order  \
count       3.243449e+07  3.243449e+07  32403719.0  ...  3.243449e+07   
mean        8.351076e+00  5.896975e-01         NaN  ...  3.305254e+01   
std         7.126671e+00  4.918886e-01         0.0  ...  2.515471e+01   
min         1.000000e+00  0.000000e+00         1.0  ...  3.000000e+00   
25%         3.000000e+00  0.000000e+00        31.0  ...  1.300000e+01   
50%         6.000000e+00  1.000000e+00        83.0  ...  2.600000e+01   
75%         1.100000e+01  1.000000e+00       107.0  ...  4.700000e+01   
max         1.450000e+02  1.000000e+00       134.0  ...  9.900000e+01   

       mean_spending  median_days_since_ordering           age  n_dependants  \
count   3.243449e+07                  32434489.0  3.243449e+07  3.243449e+07   
mean    7.029131e+00                         NaN  4.946513e+01  1.501874e+00   
std     1.020756e+00                         0.0  1.848554e+01  1.118868e+00   
min     1.000000e+00                         0.0  1.800000e+01  0.000000e+00   
25%     7.378475e+00                         5.5  3.300000e+01  1.000000e+00   
50%     7.811946e+00                         7.0  4.900000e+01  2.000000e+00   
75%     8.229365e+00                        13.0  6.500000e+01  3.000000e+00   
max     2.320000e+01                        30.0  8.100000e+01  3.000000e+00   

             income  total_spent_in_order  total_spent_in_lifetime  \
count  3.243449e+07          3.243449e+07             3.243449e+07   
mean   9.943542e+04          1.127550e+02             2.952755e+03   
std    4.305519e+04          7.210378e+01             2.752658e+03   
min    2.590300e+04          0.000000e+00             3.000000e+00   
25%    6.699800e+04          6.680000e+01             1.102200e+03   
50%    9.661200e+04          1.076000e+02             2.388400e+03   
75%    1.279080e+05          1.617000e+02             4.637100e+03   
max    5.939010e+05          1.092200e+03             2.639490e+04   

       total_items_in_lifetime  avg_items_per_order_by_user  
count             3.243449e+07                 3.243449e+07  
mean              4.224109e+02                          NaN  
std               3.788504e+02                 0.000000e+00  
min               3.000000e+00                 1.000000e+00  
25%               1.420000e+02                 8.539062e+00  
50%               3.070000e+02                 1.200000e+01  
75%               5.940000e+02                 1.657812e+01  
max               3.725000e+03                 7.025000e+01  

[8 rows x 22 columns]

## 0.4  Create new columns with profile flags

In [12]:
# Identifying the number of unique user_id numbers
df.user_id.nunique()

206209

In [13]:
# Create a subset of the df where each row = 1 user
# insted of each row being 1 item in an order
# so that the flags not involving price can be calculated by user 
# rather than by item

# Empty df
df_subset = pd.DataFrame()

In [14]:
# Subset only necessary columns: user_id, order_id, and days_since_prior
df_subset = df[['user_id', 'age', 'fam_status', 'n_dependants', 'income']]

In [15]:
# Confirming that we have the correct subset
df_subset.head()

user_id  age fam_status  n_dependants  income
0        1   31    married             3   40423
1        1   31    married             3   40423
2        1   31    married             3   40423
3        1   31    married             3   40423
4        1   31    married             3   40423

In [16]:
df_subset.user_id.nunique()

206209

In [17]:
df_subset.shape

(32434489, 5)

In [18]:
# Remove duplicate rows
df_subset = df_subset.drop_duplicates()

In [19]:
# Confirmting that the subset now has one row per user_id
df_subset.shape

(206209, 5)

In [20]:
df_subset.head()

user_id  age        fam_status  n_dependants  income
0          1   31           married             3   40423
59         2   68           married             3   64940
254        3   33           married             3   93240
342        4   31           married             3   45435
360        5   75  divorced/widowed             0  115242

### 0.4.1  income_group

In [21]:
# Exploring the income distribution in thid df
df_subset['income'].describe()

count    206209.000000
mean      94632.852548
std       42473.786988
min       25903.000000
25%       59874.000000
50%       93547.000000
75%      124244.000000
max      593901.000000
Name: income, dtype: float64

In the U.S., middle class is often defined as earning between 2/3 and twice the median American household income.  Based on 2021 information as well as a visual distribution of income in this data set done in the Data Visualizations A script, we will use the following income_group categories:

'low income' means income <50k

'middle class' means income between 50k and 150k

'upper middle class' means income between 150k and 250k

'upper class' means income >250k

In [22]:
# Creating a column 'income_group' with values based on the 'income' column:
#    If income is <50000 , then 'Low income'
#    If income is <= 150000 and >=50000, then 'Middle class'
#    If income is <= 250000 and >150000, then 'Upper middle class'
#    If income is >250000, then 'Affluent'

In [23]:
df_subset.loc[df_subset['income'] <= 50000, 'income_group'] = 'Low income'

In [24]:
df_subset.loc[df_subset['income'] > 250000, 'income_group'] = 'Affluent'

In [25]:
df_subset.loc[(df_subset['income'] <= 150000) & (df_subset['income'] > 50000), 'income_group'] = 'Middle class'

In [26]:
df_subset.loc[(df_subset['income'] <= 250000) & (df_subset['income'] > 150000), 'income_group'] = 'Upper middle class'

In [27]:
# Checking the results

df_subset['income_group'].value_counts()

Middle class          148853
Low income             34105
Upper middle class     22554
Affluent                 697
Name: income_group, dtype: int64

The sum of the counts matches the total number of expected values.

In [28]:
df_subset.head()

user_id  age        fam_status  n_dependants  income  income_group
0          1   31           married             3   40423    Low income
59         2   68           married             3   64940  Middle class
254        3   33           married             3   93240  Middle class
342        4   31           married             3   45435    Low income
360        5   75  divorced/widowed             0  115242  Middle class

In [29]:
df_subset.shape

(206209, 6)

### 0.4.2 senior_citizen

In [30]:
# Identifying the number of senior_citizens in the df
len(df_subset[df_subset['age'] >= 65])

54729

In [31]:
# Create a new column 'senior_citizen'
# If age >=65, then senior_citizen = TRUE, else FALSE

df_subset['senior_citizen'] = df_subset['age'] >= 65

In [32]:
# Confirming that the column was added correctly.
df_subset.head()

user_id  age        fam_status  n_dependants  income  income_group  \
0          1   31           married             3   40423    Low income   
59         2   68           married             3   64940  Middle class   
254        3   33           married             3   93240  Middle class   
342        4   31           married             3   45435    Low income   
360        5   75  divorced/widowed             0  115242  Middle class   

     senior_citizen  
0             False  
59             True  
254           False  
342           False  
360            True

In [33]:
df_subset['senior_citizen'].value_counts()

False    151480
True      54729
Name: senior_citizen, dtype: int64

In [34]:
df_subset.shape

(206209, 7)

### 0.4.3  two_parent_fam

In [35]:
# Create a new column 'two_parent_fam'
# If fam_status = 'married' AND n_dependants >0, 
# then two_parent_fam = TRUE, else FALSE

df_subset['two_parent_fam'] = ((df_subset['fam_status'] == 'married') & (df_subset['n_dependants'] > 0))

In [36]:
# Confirming that the column was added correctly.
df_subset.head(10)

user_id  age        fam_status  n_dependants  income        income_group  \
0          1   31           married             3   40423          Low income   
59         2   68           married             3   64940        Middle class   
254        3   33           married             3   93240        Middle class   
342        4   31           married             3   45435          Low income   
360        5   75  divorced/widowed             0  115242        Middle class   
397        6   48           married             1  116041        Middle class   
411        7   39           married             2   89095        Middle class   
617        8   32           married             1   88603        Middle class   
666        9   79           married             3  154743  Upper middle class   
742       10   34           married             2   52716        Middle class   

     senior_citizen  two_parent_fam  
0             False            True  
59             True            True  
254           False            True  
342           False            True  
360            True           False  
397           False            True  
411           False            True  
617           False            True  
666            True            True  
742           False            True

In [37]:
df_subset['two_parent_fam'].value_counts()

True     144906
False     61303
Name: two_parent_fam, dtype: int64

In [38]:
df_subset.shape

(206209, 8)

### 0.4.4  single_parent

In [39]:
# Create a new column 'single_parent'
# If n_dependants >0 AND fam_status does NOT equal 'married' 
# then single_parent = TRUE, else FALSE

df_subset['single_parent'] = ((df_subset['fam_status'] != 'married') & (df_subset['n_dependants'] > 0))

In [40]:
df_subset.head()

user_id  age        fam_status  n_dependants  income  income_group  \
0          1   31           married             3   40423    Low income   
59         2   68           married             3   64940  Middle class   
254        3   33           married             3   93240  Middle class   
342        4   31           married             3   45435    Low income   
360        5   75  divorced/widowed             0  115242  Middle class   

     senior_citizen  two_parent_fam  single_parent  
0             False            True          False  
59             True            True          False  
254           False            True          False  
342           False            True          False  
360            True           False          False

In [41]:
df_subset['single_parent'].value_counts()

False    196508
True       9701
Name: single_parent, dtype: int64

In [42]:
df_subset.shape

(206209, 9)

### 0.4.5  baby_parent

In [43]:
# Identifying the number of transaction in the babies department (18)
len(df[df['department_id'] == 18])

423794

In [44]:
# Identifying the number of transactions where the user has dependants
len(df[df['n_dependants'] >0])

24329247

In [45]:
# Add a column to identify potential baby_parents
# where department_id=18 AND n_dependants > 0, 
# then TRUE, else FALSE

df['baby_parent'] = ((df['n_dependants'] > 0) & (df['department_id'] == 18))


In [46]:
# Checking the counts of baby_parent
df['baby_parent'].value_counts()

False    32117189
True       317300
Name: baby_parent, dtype: int64

The number of true transactions for baby_parent (317,300) is less than the total number of baby transactions (423,794) probably because some baby transaction might be gifts or for other purposes.

In [47]:
# Making it so if a user has any transaction flagged as 'baby_parent' then
# the user is consistently flagged as a baby_parent across all transactions
df['baby_parent'] = (df.groupby(['user_id'])['baby_parent'].transform(np.sum))>0

In [48]:
# Checking the counts of baby_parent
df['baby_parent'].value_counts()

False    24948816
True      7485673
Name: baby_parent, dtype: int64

The number of true transactions have drastically increased, but that makes sense since parents of babies would make some transactions where they might not buy any baby items.  It may also be the case that their child has grown and is no longer a baby -- we would need data about the dates of transaction if we wanted to make the flags expire after a certain amount of time with no baby transactions.

In [49]:
df.shape

(32434489, 40)

In [50]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... n_dependants  fam_status  income  region  \
0                  1          0  ...            3     married   40423   South   
1                  2          0  ...            3     married   40423   South   
2                  3          0  ...            3     married   40423   South   
3                  4          0  ...            3     married   40423   South   
4                  5          0  ...            3     married   40423   South   

   low_activity_flag  total_spent_in_order total_spent_in_lifetime  \
0              False                  31.6              375.700012   
1              False                  31.6              375.700012   
2              False                  31.6              375.700012   
3              False                  31.6              375.700012   
4              False                  31.6              375.700012   

  total_items_in_lifetime avg_items_per_order_by_user baby_parent  
0                      59                    5.898438       False  
1                      59                    5.898438       False  
2                      59                    5.898438       False  
3                      59                    5.898438       False  
4                      59                    5.898438       False  

[5 rows x 40 columns]

### 0.4.6  pet_owner

In [51]:
# Identifying the number of transaction in the pet department (8)
len(df[df['department_id'] == 8])

97716

In [52]:
# Create a temporary column 'pet_transaction' to identify these transactions
# If department_id=8, then pet_transaction = TRUE, else FALSE

df['pet_transaction'] = df['department_id'] == 8

In [53]:
# Confirming the correct number of rows have been labeled as pet transactions.
df['pet_transaction'].value_counts()

False    32336773
True        97716
Name: pet_transaction, dtype: int64

In [54]:
# If the number of pet transactions for a user is greater than zero, then
# pet_owner = TRUE, else FALSE
df['pet_owner'] = (df.groupby(['user_id'])['pet_transaction'].transform(np.sum))>0

In [55]:
# Checking the counts of pet_owners
df['pet_owner'].value_counts()

False    28903972
True      3530517
Name: pet_owner, dtype: int64

In [56]:
df.shape

(32434489, 42)

In [57]:
# Removing the pet_transaction column as unnecessary now that we have the pet_owner flag
df = df.drop(columns = ['pet_transaction'])

In [58]:
df.shape

(32434489, 41)

In [59]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... fam_status  income  region  \
0                  1          0  ...    married   40423   South   
1                  2          0  ...    married   40423   South   
2                  3          0  ...    married   40423   South   
3                  4          0  ...    married   40423   South   
4                  5          0  ...    married   40423   South   

   low_activity_flag  total_spent_in_order  total_spent_in_lifetime  \
0              False                  31.6               375.700012   
1              False                  31.6               375.700012   
2              False                  31.6               375.700012   
3              False                  31.6               375.700012   
4              False                  31.6               375.700012   

  total_items_in_lifetime avg_items_per_order_by_user baby_parent pet_owner  
0                      59                    5.898438       False     False  
1                      59                    5.898438       False     False  
2                      59                    5.898438       False     False  
3                      59                    5.898438       False     False  
4                      59                    5.898438       False     False  

[5 rows x 41 columns]

### 0.5 Merging temp subset with full df

In [60]:
# Drop columns from subset that we don't want merged
# because it would result in duplicated columns
df_subset = df_subset.drop(columns = ['age', 'fam_status', 'n_dependants', 'income'])

In [61]:
df_subset.head()

user_id  income_group  senior_citizen  two_parent_fam  single_parent
0          1    Low income           False            True          False
59         2  Middle class            True            True          False
254        3  Middle class           False            True          False
342        4    Low income           False            True          False
360        5  Middle class            True           False          False

In [62]:
df = df.merge(df_subset, on = ['user_id'], how = 'left', indicator = True)

In [63]:
# Exploring the merge
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... total_spent_in_lifetime  \
0                  1          0  ...              375.700012   
1                  2          0  ...              375.700012   
2                  3          0  ...              375.700012   
3                  4          0  ...              375.700012   
4                  5          0  ...              375.700012   

   total_items_in_lifetime  avg_items_per_order_by_user  baby_parent  \
0                       59                     5.898438        False   
1                       59                     5.898438        False   
2                       59                     5.898438        False   
3                       59                     5.898438        False   
4                       59                     5.898438        False   

   pet_owner  income_group senior_citizen two_parent_fam single_parent _merge  
0      False    Low income          False           True         False   both  
1      False    Low income          False           True         False   both  
2      False    Low income          False           True         False   both  
3      False    Low income          False           True         False   both  
4      False    Low income          False           True         False   both  

[5 rows x 46 columns]

In [64]:
df.shape

(32434489, 46)

The shape is what we would expect: same number of rows, columns increased by 5 (four flags plus the merge field)

In [65]:
df['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

In [66]:
# Removing the merge flag so it doesn't interfere with future merges
df = df.drop(columns = ['_merge'])

In [67]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... total_spent_in_order  \
0                  1          0  ...                 31.6   
1                  2          0  ...                 31.6   
2                  3          0  ...                 31.6   
3                  4          0  ...                 31.6   
4                  5          0  ...                 31.6   

   total_spent_in_lifetime  total_items_in_lifetime  \
0               375.700012                       59   
1               375.700012                       59   
2               375.700012                       59   
3               375.700012                       59   
4               375.700012                       59   

   avg_items_per_order_by_user  baby_parent  pet_owner income_group  \
0                     5.898438        False      False   Low income   
1                     5.898438        False      False   Low income   
2                     5.898438        False      False   Low income   
3                     5.898438        False      False   Low income   
4                     5.898438        False      False   Low income   

  senior_citizen two_parent_fam single_parent  
0          False           True         False  
1          False           True         False  
2          False           True         False  
3          False           True         False  
4          False           True         False  

[5 rows x 45 columns]

In [68]:
df.shape

(32434489, 45)

### 0.6 Confirming Final Df

In [69]:
df.shape

(32434489, 45)

Final df has expected number of rows (32434489) and columns (37+6).

In [71]:
# Checking datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434489 entries, 0 to 32434488
Data columns (total 45 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   order_id                     int64         
 1   user_id                      int64         
 2   order_number                 int16         
 3   order_day_of_week            int8          
 4   order_hour_of_day            int8          
 5   days_since_prior_order       float16       
 6   first_order_of_new_customer  bool          
 7   product_id                   int32         
 8   add_to_cart_order            int16         
 9   reordered                    int8          
 10  product_name                 object        
 11  aisle_id                     float16       
 12  department_id                float16       
 13  price                        float32       
 14  product_info_available       bool          
 15  price_available              bool          
 16

In [72]:
# Adjust datatypes for new columns to be more appropriately sized
df['income_group'] = df['income_group'].astype('category')

In [73]:
# Confirmting datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434489 entries, 0 to 32434488
Data columns (total 45 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   order_id                     int64         
 1   user_id                      int64         
 2   order_number                 int16         
 3   order_day_of_week            int8          
 4   order_hour_of_day            int8          
 5   days_since_prior_order       float16       
 6   first_order_of_new_customer  bool          
 7   product_id                   int32         
 8   add_to_cart_order            int16         
 9   reordered                    int8          
 10  product_name                 object        
 11  aisle_id                     float16       
 12  department_id                float16       
 13  price                        float32       
 14  product_info_available       bool          
 15  price_available              bool          
 16

### 0.7 Exporting the new df

In [74]:
#Exporting final df
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all3.pkl'))